homeassistant/components/hyperion/light.py

```diff
@@ -515,9 +515,12 @@ def _set_internal_state(
                 self._icon = ICON_EXTERNAL_SOURCE
             else:
                 self._icon = ICON_EFFECT
+        if not self.is_on:
+            self._icon = ICON_LIGHTBULB
 
     def _update_components(self, _: Optional[Dict[str, Any]] = None) -> None:
         """Update Hyperion components."""
+        self._set_internal_state()
```

Comment:
Sorry to change direction, but your comments provoked me to think a bit more about this. The issue here is that it entangles the icon with the update of components (and as it happens, to illustrate the problem, I have a PR brewing that will change what `turn_on` acts upon in some circumstances, so components would then be the wrong place to call this from).

I think a better solution is to remove the self._icon variable entirely, and move the whole if structure that sets the icon from here into the `def icon` property itself (where you  originally had made your addition).

Then whenever `self.async_write_ha_state()` the icon will be re-fetched. This method is likely already called in all the right  places, so you could just update the `def icon` method and be confident the right thing will happen.

Reply:
No problem, I just made a commit to move the icon logic.